vissl/utils/events.py

```diff
@@ -0,0 +1,143 @@
+import json
+from collections import defaultdict
+
+from fvcore.common.file_io import PathManager
+from fvcore.common.history_buffer import HistoryBuffer
+
+
+_VISSL_EVENT_STORAGE_STACK = []
+
+
+def get_event_storage():
+    return _VISSL_EVENT_STORAGE_STACK[-1]
+
+
+def create_event_storage():
+    _VISSL_EVENT_STORAGE_STACK.append(VisslEventStorage())
+
+
+class VisslEventWriter:
+    """
+    Base class for writers that obtain events from :class:`VisslEventStorage`
+    and process them.
+    """
+
+    def write(self):
+        raise NotImplementedError
+
+    def close(self):
+        pass
+
+
+class VisslEventStorage:
+    """
+    The user-facing class that stores the running metrics
+    and all the logging data in one place. The Storage can
+    be updated anywhere in the training step. The storage
+    is used by several writers/hooks to write the data to
+    several backends (json, tensorboard, etc)
+    """
+
+    def __init__(self, start_iter=0):
+        """
+        Args:
+            start_iter (int): the iteration number to start with
+        """
+        self._history = defaultdict(HistoryBuffer)
+        self._latest_scalars = {}
+        self._iter = start_iter
+        # self._vis_data = []     # later for tensorboard
+        # self._histograms = []   # later for tensorboard
+
+    def put_scalar(self, name, value):
+        """
+        Add a scalar `value` to the `HistoryBuffer` associated with `name`.
+        """
+        history = self._history[name]
+        value = float(value)
+        history.update(value, self._iter)
+        self._latest_scalars[name] = (value, self._iter)
+
+    def put_scalars(self, **kwargs):
+        """
+        Put multiple scalars from keyword arguments.
+        Examples:
+            storage.put_scalars(loss=my_loss, accuracy=my_accuracy)
+        """
+        for k, v in kwargs.items():
+            self.put_scalar(k, v)
+
+    def history(self, name):
+        """
+        Returns:
+            HistoryBuffer: the scalar history for name
+        """
+        ret = self._history.get(name, None)
+        if ret is None:
+            raise KeyError(f"No history metric available for {name}!")
+        return ret
+
+    def histories(self):
+        """
+        Returns:
+            dict[name -> HistoryBuffer]: the HistoryBuffer for all scalars
+        """
+        return self._history
+
+    def latest(self):
+        """
+        Returns:
+            dict[str -> (float, int)]: mapping from the name of each scalar to the most
+                recent value and the iteration number its added.
+        """
+        return self._latest_scalars
+
+    @property
+    def iter(self):
+        """
+        Returns:
+            int: The current iteration number.
+        """
+        return self._iter
```

Comment:
insert pdb or print(self._iter)

Reply:
@prigoyal: We debugged the inconsistencies with the iteration value using the test you provided us: Please, find the log below:

```{"batch_time": 6273.0, "epoch": 0.0, "eta": 150.56117820739746, "iteration": 1, "loss": 1.3317, "lr": 0.00234, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5147.0, "epoch": 0.0, "eta": 102.94003486633301, "iteration": 4, "loss": 1.28832, "lr": 0.0022, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5190.0, "epoch": 0.0, "eta": 77.85301351547241, "iteration": 9, "loss": 1.24261, "lr": 0.00167, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5131.0, "epoch": 0.0, "eta": 51.3175384203593, "iteration": 14, "loss": 1.31571, "lr": 0.00095, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5178.0, "epoch": 0.0, "eta": 25.892525255680084, "iteration": 19, "loss": 0.75789, "lr": 0.00032, "peak_mem_used": -1.0, "rank": 0.0}
{"batch_time": 5142.0, "epoch": 0.0, "eta": 0.0, "iteration": 24, "loss": 1.28264, "lr": 1e-05, "peak_mem_used": -1.0, "rank": 0.0}```

Reply:
can you share the log.txt file ? how many iterations ran  and does log.txt show updated epochs number? also where we are logging the epochs in this change, can you insert pdb and verify just like we did debugging for "iteration" ?